### ISSUE TITLE
Impossible to move elements from different pools

### ISSUE THREAD

**Issue by pedesen** [bpmn-io/bpmn-js#524]
Via [this forum topic](https://forum.camunda.org/t/dragging-task-boxes-in-different-pool-participants/518).

![2905445186f3087bcaff8937e59e87b5539701a6](https://cloud.githubusercontent.com/assets/2543743/14852443/9207adca-0c86-11e6-9172-e03f01796907.gif)

__Steps to reproduce__

* Open [`diagram.bpmn.txt`](https://github.com/bpmn-io/bpmn-js/files/340985/diagram.bpmn.txt)
* Select `Task_A`, `Task_B` and the connecting flow
* Try to move all selected elements

__Problem__

* The move operation is prevented by rules

__Expected behavior__

* It should be possible to move all elements as long as the primary selection (the element behind the cursor) does not change its parent.

__Hints___

* The [`Move` feature](https://github.com/bpmn-io/diagram-js/blob/master/lib/features/move/Move.js#L35) checks continuously during element moving whether the operation is allowed.

**Comment by jocutajar** [comment]
Hi,

I have [tried to reproduce this issue in a test](https://github.com/jocutajar/bpmn-js/commit/cada91d54bf402c08843f8d70e0fd2602a3fe7d1), but haven't managed yet. It appears that move is always allowed. How can I get the rules defined in [BpmnRules.js](https://github.com/jocutajar/bpmn-js/blob/master/lib/features/rules/BpmnRules.js) under test?

Also, it occurred to me that if you move a selection of multiple elements in different pools, you may well check that the primary dragged element did not change parent to allow the move. Still the remaining elements may end up outside of the boundary of their parents which would be another bug. To fix this properly, you'd have to check each and every element's parent not changing in the new position. This would imply some more verbose changes in the code as the context does not immediately provide this information to the rules.

Thanks, Robert

**Comment by pedesen** [comment]
Hi Robert,

thanks for providing a solution to this issue.

Your assumption that move is always allowed in the test case is semi-true ;) In [MoveElementsSpec.js](https://github.com/jocutajar/bpmn-js/blob/cada91d54bf402c08843f8d70e0fd2602a3fe7d1/test/spec/features/modeling/MoveElementsSpec.js#L83-L106) you are moving elements via `moveElements`. In this case no bpmn rules are checked, because it is not a dragging operation.

In [BpmnRulesSpec.js](https://github.com/jocutajar/bpmn-js/blob/cada91d54bf402c08843f8d70e0fd2602a3fe7d1/test/spec/features/rules/BpmnRulesSpec.js#L900-L916) you are checking the rules directly. This is the right approach. But: You provided the wrong element id for the target. There is no `Pool_A` in your bpmn diagram. Try `Participant_1` instead and the test case will fail.

Another hint for moving non-primary elements outside its parents: This behavior should be prevented using the auto expand feature, so all elements should stay in their previous parent. Therefore the 'parent check' may happen for the primary selected element only. This already works for sub processes:

![multiselect](https://cloud.githubusercontent.com/assets/2543743/18084782/d40d0696-6ea8-11e6-8570-8b96fbb9c755.gif)

**Comment by jocutajar** [comment]
Hi Patrick,

Thanks for the hints. Finally, I got some time to continue. After changing the pool/participant name as you suggested and fixing some other issue I managed to produce a [failing test ](https://github.com/jocutajar/bpmn-js/commit/b8de084e6a2b250ff078f3e2ac08d9afe6f6d335) and then fix the problem - see [comit fcc69fa](https://github.com/jocutajar/bpmn-js/commit/fcc69fa79ef33faf30e34cb96b7a3d899bd496f5).

Before the fix: The move is prevented by MessageFlow element being included in the moved selection. In fact, when you add bends to the connection edge so that you can select only the tasks, the move proceeds (and auto expands the pool when overlapping).

Fix: I have eased the restrictions on MessageFlow canDrop rules somewhat and we pass previously failing tests. This doesn't break any other previous tests, but I'm uncertain it would be _the correct solution_.

I have a question if a move should be prevented when parents DO change. For this I have created a new failing test - '[move] is not allowed when parent does change'. Let me know if this should also be addressed or can be ignored.

Thanks and kind regards, Rob

**Comment by jocutajar** [comment]
Just thinking about it, an alternative solution would be to exclude connection (MessageFlow) elements from the moved selection as they seem to move automatically with their sources/targets. This on the precondition that both source and target elements are included in the selection.

**Comment by pedesen** [comment]
Hi Rob,

you're right, MessageFlows as part of the selection cause the problem and to ease the restriction in the canDrop rules is a valid solution. I also tried out to model this in the diagram and it seems to be fine.

Of course we could discuss, if MessageFlows should be part of a move operation or not, but this would probably part of a bigger refactoring. So for now it would go for your solution.

To your question regarding the parent change. We also discussed this internally a while ago (as we had to decide this for sub processes) and we came to the conclusion that we want to allow such a move. If you move elements to a new parent, the new target will be highlighted and all moved elements will be part of it. Personally I still think this is a good solution and a somehow predictable behavior. I don't see the need to change that.

Thanks
Patrick
